### ISSUE TITLE
Dangling symlinks in *.wants can cause dbus floods

### ISSUE THREAD

**Issue by berolinux** [systemd/systemd#16822]
**systemd version the issue has been seen with**
246 and current 246-stable

**Used distribution**
OpenMandriva Cooker

**Problem description**
While porting OpenMandriva to the PinePhone, we ran into a "funny" problem: the system would "sort-of" work, but took forever to start, with systemd (PID 1), systemd-logind and dbus-broker each taking up 100% CPU, and dbus-broker using up all available memory.

An investigation revealed PID 1 was stuck in an infinite loop executing
```
sendmsg(54, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\4\1\1|\3\0\0\275o\f\0\256\0\0\0\1\1o\0H\0\0\0/org/freedesktop/systemd1/unit/systemd_2dboot_2dsystem_2dtoken_2eservice\0\0\0\0\0\0\0\0\2\1s\0\37\0\0\0org.freedesktop.DBus.Properties\0\3\1s\0\21\0\0\0PropertiesChanged\0\0\0\0\0\0\0\10\1g\0\10sa{sv}as\0\0\0", iov_len=192}, {iov_base="\35\0\0\0org.freedesktop.systemd1.Unit\0\0\0004\3\0\0\v\0\0\0ActiveState\0\1s\0\0\10\0\0\0inactive\0\0\0\0\0\0\0\0\f\0\0\0FreezerState\0\1s\0\7\0\0\0running\0\10\0\0\0SubState\0\1s\0\4\0\0\0dead\0\0\0\0\0\0\0\0\24\0\0\0StateChangeTimestamp\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\0\35\0\0\0StateChangeTimestampMonotonic\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\25\0\0\0InactiveExitTimestamp\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\36\0\0\0InactiveExitTimestampMonotonic\0\1t\0\0\0\0\0\0\0\0\0\0\0\24\0\0\0ActiveEnterTimestamp\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\0\35\0\0\0ActiveEnterTimestampMonotonic\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\23\0\0\0ActiveExitTimestamp\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\0\0\34\0\0\0ActiveExitTimestampMonotonic\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\0\26\0\0\0InactiveEnterTimestamp\0\1t\0\0\0\0\0\0\0\0\0\0\0\37\0\0\0InactiveEnterTimestampMonotonic\0\1t\0\0\0\0\0\0\0\0\0\0\3\0\0\0Job\0\4(uo)\0\0\0\0\0\0\0\1\0\0\0/\0\0\0\0\0\0\0\17\0\0\0ConditionResult\0\1b\0\0\0\0\0\0\0\0\0\0\f\0\0\0AssertResult\0\1b\0\0\0\0\0\22\0\0\0ConditionTimestamp\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\33\0\0\0ConditionTimestampMonotonic\0\1t\0\0\0\0\0\0\0\0\0\0\0\0\0\0\17\0\0\0AssertTimestamp\0\1t\0\0\0\0\0\0\0\0\0\0\30\0\0\0AssertTimestampMonotonic\0\1t\0\0\0\0\0\0\0\0\0\f\0\0\0InvocationID\0\2ay\0\0\0\0\0\0\0\0\34\0\0\0\n\0\0\0Conditions\0\0\7\0\0\0Asserts\0", iov_len=892}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 1084
newfstatat(AT_FDCWD, "/etc/systemd/system", {st_mode=S_IFDIR|0755, st_size=3488, ...}, 0) = 0
newfstatat(AT_FDCWD, "/etc/systemd/system.attached", 0x7ff6792840, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/run/systemd/system", {st_mode=S_IFDIR|0755, st_size=40, ...}, 0) = 0
newfstatat(AT_FDCWD, "/run/systemd/system.attached", 0x7ff6792840, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/local/lib/systemd/system", 0x7ff6792840, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/lib/systemd/system", {st_mode=S_IFDIR|0755, st_size=20480, ...}, 0) = 0
newfstatat(AT_FDCWD, "/usr/lib/systemd/system", 0x7ff6792840, 0) = -1 ENOENT (No such file or directory)
clock_gettime(CLOCK_REALTIME, {tv_sec=1597682466, tv_nsec=342686836}) = 0
```
over and over.

It turns out this was caused by forgetting to remove the /lib/systemd/system/sysinit.target.wants/systemd-boot-system-token.service symlink (pointing at ../systemd-boot-symlink-token.service, which didn't exist - PinePhone doesn't use UEFI or systemd-boot).

This must have caused something to go into an infinite loop requesting the service start over D-Bus, in turn flooding the system bus, making everything using it (such as systemd-logind) time out.

I can not reproduce this by creating a dangling symlink on in sysinit.target.wants on my desktop; so there must be an extra condition required to trigger this (possibly the lack of an initramfs/initrd?).

If this can be triggered for user sessions as well, there might be a way to trigger a DOS.

**Comment by lethalwp** [comment]
this is also a bug present in current pre-beta fedora 33, which uses systemd-246.1.

if there's a "wanted" target file left on the system from a previous fedora release, not removed after an update, this bug is triggered.

It is also triggered when a 'want' directory is empty.

This causes dbus to disable user 0, which also kills gdm in fedora 33-pre-beta

**Comment by lethalwp** [comment]
looks to be fixed in the current fedora rawhide (246.4)

**Comment by keszybz** [comment]
Thanks for testing, let's close.
